### PR TITLE
Add glu op

### PR DESCRIPTION
**Pull request by beverlylytle**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Adds support for the glu op

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: thunder/extend/__init__.py
Comment (crcrpar) [inline thunder/extend/__init__.py:1]: Is this a change of import order and cleanup of unused imports?

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -1345,7 +1345,7 @@ def view_as(a: TensorLike, b: TensorLike, /) -> TensorLike:
 
 
 #
-# Elementwise unary operaitons
+# Elementwise unary operations
```
Comment (crcrpar) [inline thunder/torch/__init__.py:1348]: good catch

File: thunder/torch/__init__.py [hunk 2]
```diff
@@ -2687,6 +2687,19 @@ def clone(a: TensorProxy, *, memory_format=torch.preserve_format) -> TensorProxy
 register_method("clone", clone)
 
 
+@torchsymbol(torch.nn.functional.glu, is_method=False)
+def glu(a: TensorProxy, /, dim: int = -1):
+    dim = -1 if dim is None else dim
```
Comment (crcrpar) [inline thunder/torch/__init__.py:2692]: When would `dim` be `None`? The default is -1 so I wasn't sure at glance.
It might be better to make sure `0 <= dim < a.ndim` or `dim == -1`

File: thunder/extend/__init__.py
Comment (beverlylytle) [reply thunder/extend/__init__.py:1]: Yes.

**Review by crcrpar — APPROVED** [review]
this LGTM

File: thunder/tests/opinfos.py [hunk 3]
```diff
@@ -5491,6 +5488,54 @@ def var_sample_generator(op, device, dtype, requires_grad):
...
+    make = partial(
+        make_tensor,
+        device=device,
+        dtype=dtype,
+        requires_grad=requires_grad,
+    )
+
+    cases = (
+        ((4,),),
+        ((3, 4), 1),
```
Comment (mruberry) [inline thunder/tests/opinfos.py:5501]: What about a dim of -1?

File: thunder/tests/opinfos.py [hunk 4]
```diff
@@ -5491,6 +5488,54 @@ def var_sample_generator(op, device, dtype, requires_grad):
...
+    yield (SampleInput(make((3,))), RuntimeError, err_msg)
+    yield (SampleInput(make((2, 2, 3))), RuntimeError, err_msg)
+    yield (SampleInput(make((4, 5, 8)), dim=1), RuntimeError, err_msg)
+
+
+glu_opinfo = OpInfo(
+    ltorch.glu,
+    sample_input_generator=glu_sample_generator,
+    error_input_generator=glu_error_generator,
+    dtypes=(datatypes.floating, datatypes.complexfloating),
```
Comment (mruberry) [inline thunder/tests/opinfos.py:5532]: nit: `datatypes.inexact`?

File: thunder/torch/__init__.py (hunk 1 above)
Comment (mruberry) [inline thunder/torch/__init__.py:1348]: Nice spelling fix

File: thunder/torch/__init__.py [hunk 5]
```diff
@@ -2687,6 +2687,23 @@ def clone(a: TensorProxy, *, memory_format=torch.preserve_format) -> TensorProxy
 register_method("clone", clone)
 
 
+@torchsymbol(torch.nn.functional.glu, is_method=False)
+def glu(a: TensorProxy, /, dim: int = -1):
+    utils.check(
```
Comment (mruberry) [inline thunder/torch/__init__.py:2692]: I don't think this needs to check for the dim being within the expected range, but if you'd like to then maybe just call `canonicalize_dim`?

https://github.com/Lightning-AI/lightning-thunder/blob/15c48efc7fa2befd628ffa0e86a3e2d17816b9de/thunder/core/utils.py#L585

File: thunder/torch/__init__.py [hunk 6]
```diff
@@ -2687,6 +2687,23 @@ def clone(a: TensorProxy, *, memory_format=torch.preserve_format) -> TensorProxy
...
+        -a.ndim <= dim < a.ndim,
+        lambda: f"Dimension out of range (expected to be in range [{-a.ndim}, {a.ndim - 1}], but got {dim})",
+        IndexError,
+    )
+    utils.check(
+        a.shape[dim] % 2 == 0,
+        lambda: f"Halving dimension must be even, but dimension {dim} is size {a.shape[dim]}",
+    )
+    chunk_size = a.shape[dim] // 2
+    left, right = split(a, (chunk_size, chunk_size), dim=dim)
```
Comment (mruberry) [inline thunder/torch/__init__.py:2702]: Nice implementation

**Review by mruberry — COMMENTED** [review]
I just made a couple of small suggestions. Looks great overall!

File: thunder/tests/opinfos.py (hunk 3 above)
Comment (beverlylytle) [reply thunder/tests/opinfos.py:5502]: added

File: thunder/torch/__init__.py [hunk 7]
```diff
@@ -2705,7 +2701,7 @@ def glu(a: TensorProxy, /, dim: int = -1):
 
 
 @torchsymbol(torch.mean, is_method=True)
-def mean(a: TensorProxy, /, dim=None, keepdim: bool = False, *, dtype=None):
+def mean(a: TensorProxy, /, dim=None, keepdim: bool = False, *, dtype=None) -> TensorProxy:
```
Comment (mruberry) [inline thunder/torch/__init__.py:2704]: Nice bonus typing fix

**Comment by mruberry** [comment]
fyi @t-vi Test failure is unrelated. I filed https://github.com/Lightning-AI/lightning-thunder/issues/1514 for it
